### PR TITLE
contractcourt: keep channels with timed-out initiated htlcs.

### PR DESCRIPTION
**Pull request by roeierez**
This PR main goal is to reduce force closed channels in a mobile node.
In case of an outgoing pending htlc, a forwarding node needs to make sure it is either cancelled or resolved on-chain after timeout.
While for an online node it is more likely that its counterpart peer will send `update_fail_htlc` before it will need to use the timeout transaction, in mobile node it is a bit different.
By its nature, mobile node is most of the time not online, which in case of pending htlc it is likely to miss the time slot for getting the `update_fail_htlc` and when comes online it will close the channel on-chain.
While in a forwarding payment the timeout transaction should be used immediately to eliminate the risk in loosing funds, it looks to me that in an initiated payment it is not risky to keep the channel at this situation. The advantage is that when the mobile node is online it will have a chance to receive the `udpate_fail_htlc`, preventing the channel from being closed.
This behavior also makes paying hold invoices in mobile a lot more practical. Assuming the hold invoice is waiting for some external info to be resolved, it can take time in which the mobile goes offline.
If the counterpart peer misbehaves and doesn't send the `udpate_fail_htlc`, then the user can always force close the channel by himself.

A flag was added to the configuration to be able to explicitly specify this behavior and in such case for every outgoing htlc that is closed to timeout and beyond the code doesn't take on-chain action if the htlc is associated with a direct payment of the user node.

### CODE REVIEW

**Comment by Roasbeef** [comment]
If a node never goes on chain for timed out HTLCs (ever), then the funds will be suspended until the channel is closed. Even then, you cannot perform a cooperative channel close while you have outstanding outgoing HTLCs. As a result, the channel is locked into force closing at _some time_ in the future once it holds onto a timed out HTLC. It's also the case that as more HTLCs build up on the commitment transaction, it becomes more expensive to completely sweep all the funds. 

Rather than _never_ going on chain to time out HTLCs, a more flexible path would be following through with this issue: https://github.com/lightningnetwork/lnd/issues/1226. The goal here is to factor additional context other than the expiry of an HTLC when deciding to go to chain. In this case, rather than never going to chain, we could only go to chain after some % threshold of funds become unavailable.

**Comment by roeierez** [comment]
Thanks @Roasbeef  for going over this and for all the details you added. 
I believe https://github.com/lightningnetwork/lnd/issues/1226 handles the case where the other peer didn't behave well by not cancelling the htlc on time or maybe it wasn't online to do so.

> In this case, rather than never going to chain, we could only go to chain after some % threshold of funds become unavailable.

The intention of this PR is to give a chance for the other peer to cancel (because the user node was offline) rather than never going on-chain. I think if the user node won't get the cancel then https://github.com/lightningnetwork/lnd/issues/1226 comes to play.

**Comment by joostjager** [comment]
It could be that the outgoing payment serves to learn a preimage that is needed to settle an on-chain contract, like for example in the case of loop out.

If we wouldn't go to chain, the on-chain contract may expire and thereafter the off-chain payment can still be settled.

Isn't there a less impactful solution possible? A way to give peers some time to send their updates, but with a timeout after which we still go to chain.

**Comment by roeierez** [comment]
> If we wouldn't go to chain, the on-chain contract may expire and thereafter the off-chain payment can still be settled.

A good point.

> Isn't there a less impactful solution possible? A way to give peers some time to send their updates, but with a timeout after which we still go to chain.

Agree with the approach.
One way to lower the impact is to record the start time when creating the `ChannelArbitrator` and then on `shouldGoOnChain`, that is called for every new block, check that a certain timeout has passed (1 minute as default should be enough) before going to chain.
During this time it is expected that both peers will be connected and the update htlc will be received while failing to do so will result in going to chain.
I will push a change at this direction looking for your feedback.

**Comment by roeierez** [comment]
@joostjager added a grace period before force closing the channel.

**Comment by joostjager** [comment]
Ah, indeed, loop in. For loop out, the payer knows the preimage already

File: contractcourt/channel_arbitrator.go [hunk 1]
```diff
@@ -1114,8 +1114,28 @@ func (c ChainActionMap) Merge(actions ChainActionMap) {
...
-
+	currentHeight uint32, rHash [32]byte) (bool, error) {
+
+	// In the case where the user specified to not force close channels that
+	// have outgoing htlc which is a result of its own payment, then we execute
+	// this check and try to exit early.
+	// This shouldn't add any security risk as there is no incoming htlc to
+	// fulfill at this case and the expectation is that when the channel is
+	// is active the other node will send update_fail_htlc to remove the htlc
+	// without closing the channel. it is up to the user to force close the
```
Comment (cfromknecht) [inline contractcourt/channel_arbitrator.go]: it -> It

File: contractcourt/channel_arbitrator.go [hunk 2]
```diff
@@ -1114,8 +1114,28 @@ func (c ChainActionMap) Merge(actions ChainActionMap) {
 // last minute as we want to ensure that when we *need* (HTLC is timed out) to
 // sweep, the commitment is already confirmed.
 func (c *ChannelArbitrator) shouldGoOnChain(htlcExpiry, broadcastDelta,
-	currentHeight uint32) bool {
-
+	currentHeight uint32, rHash [32]byte) (bool, error) {
+
+	// In the case where the user specified to not force close channels that
```
Comment (cfromknecht) [inline contractcourt/channel_arbitrator.go]: nice comment!

File: contractcourt/chain_arbitrator.go [hunk 3]
```diff
@@ -154,10 +155,10 @@ type ChainArbitratorConfig struct {
 	// resolution.
 	OnionProcessor OnionProcessor
 
-	// KeepChannelsWithPendingPayments indicates if we will force close
-	// channels with outgoing htlcs that have timed-out and they are
-	// initiated by this node (e.g initiated payment).
-	KeepChannelsWithPendingPayments bool
+	// PaymentsExpirationGracePriod indicates is a time window we let the
+	// other node to cancel an ougoing htlc that our node has initiated and
```
Comment (cfromknecht) [inline contractcourt/chain_arbitrator.go]: ougoing -> outgoing

File: contractcourt/channel_arbitrator.go [hunk 4]
```diff
@@ -1116,22 +1121,25 @@ func (c ChainActionMap) Merge(actions ChainActionMap) {
...
-		}
-		if initiated {
+	initiated, err := c.cfg.IsInitiatedPayment(rHash)
+	if err != nil {
+		return false, err
+	}
+	if initiated {
+		runningDuration := time.Now().Sub(c.startTimestamp)
+		if runningDuration < c.cfg.PaymentsExpirationGracePriod ||
+			currentHeight < htlcExpiry {
```
Comment (cfromknecht) [inline contractcourt/channel_arbitrator.go]: should this case cause the following log statement? perhaps it should be checked above with `initiated`? not totally clear where this was checked made before and why it's added here.

Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: This case is actually for the return statement following the log statement.
Anyway, looking again I realized there is a more intuitive way to implement this case that is also easier to understand as an increment to the old code.
Pushed the changes.

File: contractcourt/chain_arbitrator.go (hunk 3 above)
Comment (roeierez) [reply contractcourt/chain_arbitrator.go]: fixed.

File: contractcourt/channel_arbitrator.go (hunk 1 above)
Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: fixed

File: contractcourt/channel_arbitrator.go [hunk 5]
```diff
@@ -1194,10 +1228,13 @@ func (c *ChannelArbitrator) checkCommitChainActions(height uint32,
 			continue
 		}
 
-		toChain := c.shouldGoOnChain(
+		toChain, err := c.shouldGoOnChain(
 			htlc.RefundTimeout, c.cfg.IncomingBroadcastDelta,
-			height,
+			height, htlc.RHash,
```
Comment (joostjager) [inline contractcourt/channel_arbitrator.go]: Could it be worth to split `shouldGoOnChain` in two versions, one for incoming and one for outgoing? Then we also don't need to pass in the hash and do the payment lookup.

File: server.go [hunk 6]
```diff
@@ -915,11 +915,19 @@ func newServer(listenAddrs []net.Addr, chanDB *channeldb.DB,
...
-		NotifyClosedChannel: s.channelNotifier.NotifyClosedChannelEvent,
-		OnionProcessor:      s.sphinx,
+		DisableChannel:               s.chanStatusMgr.RequestDisable,
+		Sweeper:                      s.sweeper,
+		Registry:                     s.invoices,
+		NotifyClosedChannel:          s.channelNotifier.NotifyClosedChannelEvent,
+		OnionProcessor:               s.sphinx,
+		PaymentsExpirationGracePriod: cfg.PaymentsExpirationGracePriod,
+		IsInitiatedPayment: func(hash [32]byte) (bool, error) {
+			payment, err := paymentControl.FetchPayment(hash)
```
Comment (joostjager) [inline server.go]: This looks like a layering violation. Payments are a higher level concept and can get complicated with mpp (multiple outgoing htlcs with the same hash) and amp (multiple htlcs with different hashes). Maybe it is better to check the circuit map to see if this is a forward?

File: contractcourt/channel_arbitrator.go (hunk 5 above)
Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: > Could it be worth to split `shouldGoOnChain` in two versions

I will give it a try and see if it is more clear this way.


> Then we also don't need to pass in the hash and do the payment lookup

We will need the lookup anyway to differentiate forward from initiated payment.

Comment (joostjager) [reply contractcourt/channel_arbitrator.go]: Isn't an incoming htlc never an initiated payment?

Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: > Isn't an incoming htlc never an initiated payment?

Indeed, it wasn't clear to me that you intended to avoid the lookup only for the incoming. Understood now.

File: server.go (hunk 6 above)
Comment (roeierez) [reply server.go]: I agree. I think instead we can query the Switch for whether this is a forward.
@joostjager  Is it correct to say that if the field `AddRef` of `PaymentCircuit` struct is **NOT** nil then we are dealing with a forward?

Comment (joostjager) [reply server.go]: I think we mostly check `incomingChanID == hop.Source`

File: contractcourt/channel_arbitrator.go (hunk 5 above)
Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: I passed the htlc struct to the `shouldGoToChain` method so we can take a decision there separately for incoming/outgoing htlcs and avoid the query in case of incoming.

File: server.go (hunk 6 above)
Comment (roeierez) [reply server.go]: @joostjager  The implementation of the check for forward/initiated htlc is now moved to the switch.

File: htlcswitch/switch.go [hunk 7]
```diff
@@ -460,6 +460,18 @@ func (s *Switch) UpdateForwardingPolicies(
 	s.indexMtx.RUnlock()
 }
 
+// HasForwardedHTLCWithHash checks for a given htlc hash if it is related
+// to an existing circuit that represents a forwarded payment.
+func (s *Switch) HasForwardedHTLCWithHash(paymentHash [32]byte) bool {
+	paymentCircuites := s.circuits.LookupByPaymentHash(paymentHash)
```
Comment (joostjager) [inline htlcswitch/switch.go]: Circuites -> circuits

File: contractcourt/channel_arbitrator.go [hunk 8]
```diff
@@ -1132,7 +1137,29 @@ func (c *ChannelArbitrator) shouldGoOnChain(htlcExpiry, broadcastDelta,
...
+	// both nodes to connect and give a chance to the other node to send its
+	// updates and cancel the htlc.
+	// This shouldn't add any security risk as there is no incoming htlc to
+	// fulfill at this case and the expectation is that when the channel is
+	// active the other node will send update_fail_htlc to remove the htlc
+	// without closing the channel. It is up to the user to force close the
+	// channel if the peer misbehaves and doesn't send the update_fail_htlc.
+	// It is useful when this node is most of the time not online and is
+	// likely to miss the time slot where the htlc may be cancelled.
+	hasForward := c.cfg.HasForwardedHTLCWithHash(htlc.RHash)
```
Comment (joostjager) [inline contractcourt/channel_arbitrator.go]: We know the channel id here and the htlc index, so it could be a lookup by outgoing circuit instead of looking up all circuits with that hash.

Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: I see now you probably mean to use `LookupOpenCircuit` to fetch the exact circuit.
To execute this query I need to build the `CirtuitKey`. Quick question: Is the `HtlcIndex` of the `channeldb.HTLC` represents the `HtlcID` of the `CirtuitKey` ?

Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: I pushed changes with this feedback.

Comment (joostjager) [reply contractcourt/channel_arbitrator.go]: Ok, looks good. Yes, `HtlcIndex` is `HtlcId`.

**Comment by joostjager** [comment]
Implementation LGTM, but it would be good to have some unit test coverage on this. In order to do this, the call to `time.Now` can be replaced by injecting a `clock.Clock` dependency.

**Comment by roeierez** [comment]
> it would be good to have some unit test coverage on this

I Wanted to see that the changes are accepted before moving on to this. I will definitely work on that now.

**Comment by roeierez** [comment]
Test added.

File: contractcourt/channel_arbitrator_test.go [hunk 9]
```diff
@@ -1811,3 +1818,87 @@ func TestChannelArbitratorDanglingCommitForceClose(t *testing.T) {
...
+		resolvers: make(map[ContractResolver]struct{}),
+	}
+	chanArbCtx, err := createTestChannelArbitrator(t, log)
+	if err != nil {
+		t.Fatalf("unable to create ChannelArbitrator: %v", err)
+	}
+	chanArb := chanArbCtx.chanArb
+
+	// We'll inject a test clock implementation so we can control the uptime.
+	startTime := time.Now()
```
Comment (joostjager) [inline contractcourt/channel_arbitrator_test.go]: For a deterministic test, it is better to inject a constant time here.

File: contractcourt/channel_arbitrator_test.go [hunk 10]
```diff
@@ -1811,3 +1818,87 @@ func TestChannelArbitratorDanglingCommitForceClose(t *testing.T) {
...
+
+	// We will advance the uptime to 10 seconds which should be still within
+	// the grace period and should not trigger going to chain.
+	testClock.SetTime(startTime.Add(time.Second * 10))
+	chanArbCtx.blockEpochs <- &chainntnfs.BlockEpoch{Height: 5}
+	chanArbCtx.AssertState(StateDefault)
+
+	// We will advance the uptime to 16 seconds which should trigger going
+	// to chain.
+	testClock.SetTime(startTime.Add(time.Second * 16))
```
Comment (joostjager) [inline contractcourt/channel_arbitrator_test.go]: Nice unit test :+1:

File: htlcswitch/switch.go [hunk 11]
```diff
@@ -460,6 +460,18 @@ func (s *Switch) UpdateForwardingPolicies(
...
+// to an opened circuit that represents a forwarded payment.
+func (s *Switch) IsForwardedHTLC(chanID lnwire.ShortChannelID,
+	htlcIndex uint64) bool {
+
+	circuit := s.circuits.LookupOpenCircuit(channeldb.CircuitKey{
+		ChanID: chanID,
+		HtlcID: htlcIndex,
+	})
+	return circuit != nil && circuit.Incoming.ChanID != hop.Source
+}
```
Comment (joostjager) [inline htlcswitch/switch.go]: This function isn't covered by the test above, but is important for proper behavior of the mechanism introduced in this pr. A low-effort option could be to add a few asserts in the existing switch tests.

File: contractcourt/channel_arbitrator.go [hunk 12]
```diff
@@ -1113,17 +1118,17 @@ func (c ChainActionMap) Merge(actions ChainActionMap) {
 // we should go on chain to claim.  We do this rather than waiting up until the
 // last minute as we want to ensure that when we *need* (HTLC is timed out) to
 // sweep, the commitment is already confirmed.
-func (c *ChannelArbitrator) shouldGoOnChain(htlcExpiry, broadcastDelta,
+func (c *ChannelArbitrator) shouldGoOnChain(htlc channeldb.HTLC, broadcastDelta,
```
Comment (joostjager) [inline contractcourt/channel_arbitrator.go]: For ease of reviewing, the non-functional refactorings (introducing `Clock` and the function signature change of `shouldGoOnChain`) could be isolated in a separate commit.

File: contractcourt/channel_arbitrator.go [hunk 13]
```diff
@@ -1132,7 +1137,29 @@ func (c *ChannelArbitrator) shouldGoOnChain(htlcExpiry, broadcastDelta,
...
+	// both nodes to connect and give a chance to the other node to send its
+	// updates and cancel the htlc.
+	// This shouldn't add any security risk as there is no incoming htlc to
+	// fulfill at this case and the expectation is that when the channel is
+	// active the other node will send update_fail_htlc to remove the htlc
+	// without closing the channel. It is up to the user to force close the
+	// channel if the peer misbehaves and doesn't send the update_fail_htlc.
+	// It is useful when this node is most of the time not online and is
+	// likely to miss the time slot where the htlc may be cancelled.
+	isForwarded := c.cfg.IsForwardedHTLC(c.cfg.ShortChanID, htlc.HtlcIndex)
```
Comment (joostjager) [inline contractcourt/channel_arbitrator.go]: Is there a possibility of any unexpected behavior because of concurrency / race conditions? That the switch already modified the circuit when we get here?

Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: If he htlc is settled/cancelled just before this line is executed and therefore no longer exist in the switch the `isForwarded` will be `false` which will cause going on-chain after the grace period.
Such modification by the switch may as well happen before this code change and in that case  will cause going to chain immediately. 
I don't find this difference as problematic.

File: contractcourt/channel_arbitrator.go (hunk 12 above)
Comment (roeierez) [reply contractcourt/channel_arbitrator.go]: @joostjager Let me verify I understood it correct:
On top of the original code of this file:
1. First commit for the signature change only.
2. Second commit for the shouldGoOnChain body and introducing Clock (As `Clock` is used in the new implementation).

Was that the suggestion?

Comment (joostjager) [reply contractcourt/channel_arbitrator.go]: Ah yes, the `Clock` wasn't already there. In that case it is indeed only the signature change in a separate commit. But up to you whether you think it is worth it, it is non-blocking.

File: contractcourt/channel_arbitrator.go (hunk 13 above)
Comment (joostjager) [reply contractcourt/channel_arbitrator.go]: Make sense. What happens in link and what happens in the arbitrator is indeed already desynchronized.

File: htlcswitch/switch.go (hunk 11 above)
Comment (roeierez) [reply htlcswitch/switch.go]: I took the low effort path which seems to me also the right one.

Comment (joostjager) [reply htlcswitch/switch.go]: Yes, very compact :+1:

File: config.go [hunk 14]
```diff
@@ -416,15 +418,16 @@ func loadConfig() (*config, error) {
...
-		TrickleDelay:             defaultTrickleDelay,
-		ChanStatusSampleInterval: defaultChanStatusSampleInterval,
-		ChanEnableTimeout:        defaultChanEnableTimeout,
-		ChanDisableTimeout:       defaultChanDisableTimeout,
-		Alias:                    defaultAlias,
-		Color:                    defaultColor,
-		MinChanSize:              int64(minChanFundingSize),
-		NumGraphSyncPeers:        defaultMinPeers,
-		HistoricalSyncInterval:   discovery.DefaultHistoricalSyncInterval,
+		PaymentsExpirationGracePriod: defaultPaymentsExpirationGracePriod,
```
Comment (Roasbeef) [inline config.go]: PaymentsExpirationGracePriod -> PaymentsExpirationGrace _Period_ 

Propagates to several other areas in this PR.

Comment (roeierez) [reply config.go]: Thanks for noticing. Fixed.

File: contractcourt/chain_arbitrator.go [hunk 15]
```diff
@@ -153,6 +155,19 @@ type ChainArbitratorConfig struct {
...
+	// PaymentsExpirationGracePeriod indicates is a time window we let the
+	// other node to cancel an outgoing htlc that our node has initiated and
+	// has timed out.
+	PaymentsExpirationGracePeriod time.Duration
+
+	// IsForwardedHTLC checks for a given htlc, identified by channel id and
+	// htlcIndex, if it is a forwarded one.
+	IsForwardedHTLC func(chanID lnwire.ShortChannelID, htlcIndex uint64) bool
+
+	// clock is the clock implementation that ChannelArbitrator uses.
```
[the remark this answers is not in the document]
Comment (roeierez) [reply contractcourt/chain_arbitrator.go]: Thanks @cfromknecht, spelling fixed.
